File: acwing/算法提高课/第一章动态规划/1012.友好城市/1012.友好城市.py
N = 5000 + 10

f = [0] * N
a = []

n = int(input())

for _ in range(n):
    a.append(list(map(int, input().split())))

a.sort(key=lambda x: x[0])

res = 0
for i in range(n):
    f[i] = 1
    for j in range(i):
        if a[i][1] > a[j][1]:
            f[i] = max(f[i], f[j] + 1)
            res = max(res, f[i])
print(res)

File: acwing/算法提高课/第一章动态规划/1012.友好城市/1012.友好城市.cpp
#include <iostream>
#include <cstring>
#include <algorithm>

using namespace std;

typedef pair<int, int> PII;

const int N = 5010;

int f[N];
PII a[N];

int main()
{
    int n;
    cin >> n;
    for (int i = 0; i < n; i++)
        cin >> a[i].first >> a[i].second;

    sort(a, a + n);

    int res = 0;
    for (int i = 0; i < n; i++)
    {
        f[i] = 1;
        for (int j = 0; j < i; j++)
        {
            if (a[i].second > a[j].second)
            {
                f[i] = max(f[i], f[j] + 1);
                res = max(res, f[i]);
            }
        }
    }
    printf("%d\n", res);

    return 0;
}